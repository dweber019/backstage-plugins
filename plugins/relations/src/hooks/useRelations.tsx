import { configApiRef, useApi } from '@backstage/core-plugin-api';
import { ALL_RELATION_PAIRS } from '@backstage/plugin-catalog-graph';
import { useEntity } from '@backstage/plugin-catalog-react';
import { Entity } from '@backstage/catalog-model';

export interface RelationsWithLabel {
  name: string;
  label?: string;
}

export function useRelations(relations?: RelationsWithLabel[]) {
  const configApi = useApi(configApiRef);
  const { entity } = useEntity<Entity>();

  const relationsConfig = configApi.getConfigArray(
    'relationsProcessor.relations',
  );
  const relationsWithoutDefaults = relationsConfig
    .map(relationConfig =>
      relationConfig
        .getConfigArray('pairs')
        .map(
          pair =>
            [pair.getString('incoming'), pair.getString('outgoing')] as [
              string,
              string,
            ],
        )
        .filter(
          pair =>
            !ALL_RELATION_PAIRS.some(
              standardPair =>
                (standardPair[0] === pair[0] && standardPair[1] === pair[1]) ||
                (standardPair[0] === pair[1] && standardPair[1] === pair[0]),
            ),
        )
        .flat(),
    )
    .flat();

  const map = new Map<string, string[]>();
  entity.relations
    ?.filter(relation => {
      if (relations) {
        return relations.some(s => s.name === relation.type);
      }
      return relationsWithoutDefaults.includes(relation.type);
    })
    .forEach(relation => {
      if (map.has(relation.type)) {
        map.set(relation.type, [
          ...map.get(relation.type)!,
          relation.targetRef,
        ]);
      } else {
        map.set(relation.type, [relation.targetRef]);
      }
    });

  return Array.from(map.keys()).map(relationName => {
    const relationWithLabel = relations?.find(
      relation => relation.name === relationName,
    );
    return {
      label: (relationWithLabel && relationWithLabel.label) || relationName,
      entityRefs: map.get(relationName) || [],
    };
  });
}
