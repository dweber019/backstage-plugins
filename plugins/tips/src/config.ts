import { createApiRef, IdentityApi } from '@backstage/core-plugin-api';
import React from 'react';
import { Entity } from '@backstage/catalog-model';

export const tipsConfigRef = createApiRef<TipsConfig>({
  id: 'plugin.tips.config',
});

/** @public */
export interface TipsConfig {
  tips: Tip[];
}

/** @public */
export interface Tip {
  content: string | React.ReactElement;
  title: string;
  activate: (options: { entity?: Entity, identity?: IdentityApi }) => Promise<boolean>;
}
