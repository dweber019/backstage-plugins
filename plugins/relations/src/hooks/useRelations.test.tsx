import { configApiRef } from '@backstage/core-plugin-api';
import { renderHook } from '@testing-library/react';
import { ConfigReader } from '@backstage/config';
import { useRelations } from './useRelations';
import { TestApiProvider } from '@backstage/test-utils';
import { Entity } from '@backstage/catalog-model';
import { EntityProvider } from '@backstage/plugin-catalog-react';

describe('useRelationsCatalogGraphDefaults', () => {
  const config = new ConfigReader({
    relationsProcessor: {
      relations: [
        {
          sourceKind: 'Component',
          attribute: 'test',
          pairs: [
            {
              incoming: 'testOf',
              outgoing: 'testBy',
            },
          ],
        },
        {
          sourceKind: 'Component',
          attribute: 'other',
          pairs: [
            {
              incoming: 'otherOf',
              outgoing: 'otherBy',
            },
          ],
        },
        {
          sourceKind: 'Component',
          attribute: 'owner',
          pairs: [
            {
              incoming: 'ownerOf',
              outgoing: 'ownedBy',
            },
          ],
        },
      ],
    },
  });
  const entity = {
    kind: 'Component',
    metadata: {
      name: 'component-1',
      namespace: 'default',
    },
    relations: [
      {
        type: 'testOf',
        targetRef: 'test:default/test-1',
      },
      {
        type: 'otherOf',
        targetRef: 'other:default/other-1',
      },
      {
        type: 'ownerOf',
        targetRef: 'user:default/user-1',
      },
    ],
  } as unknown as Entity;

  const AllTheProviders = ({ children }: any) => {
    return (
      <TestApiProvider apis={[[configApiRef, config]]}>
        <EntityProvider entity={entity}>{children}</EntityProvider>
      </TestApiProvider>
    );
  };

  test('should return relations with relation name as label', () => {
    const { result } = renderHook(() => useRelations(), {
      wrapper: AllTheProviders,
    });

    expect(result.current).toMatchSnapshot();
  });

  test('should return relations with custom label', () => {
    const relationLabels = [
      {
        name: 'testOf',
        label: 'Custom label',
      },
    ];
    const { result } = renderHook(() => useRelations(relationLabels), {
      wrapper: AllTheProviders,
    });

    expect(result.current).toMatchSnapshot();
  });

  test('should return relations whitelisted', () => {
    const relationWhitelist = [
      {
        name: 'otherOf',
      },
    ];
    const { result } = renderHook(() => useRelations(relationWhitelist), {
      wrapper: AllTheProviders,
    });

    expect(result.current).toMatchSnapshot();
  });
});
