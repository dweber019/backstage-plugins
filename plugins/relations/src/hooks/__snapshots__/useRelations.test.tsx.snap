// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useRelationsCatalogGraphDefaults should return relations whitelisted 1`] = `
[
  {
    "entityRefs": [
      "other:default/other-1",
    ],
    "label": "otherOf",
  },
]
`;

exports[`useRelationsCatalogGraphDefaults should return relations with custom label 1`] = `
[
  {
    "entityRefs": [
      "test:default/test-1",
    ],
    "label": "Custom label",
  },
]
`;

exports[`useRelationsCatalogGraphDefaults should return relations with relation name as label 1`] = `
[
  {
    "entityRefs": [
      "test:default/test-1",
    ],
    "label": "testOf",
  },
  {
    "entityRefs": [
      "other:default/other-1",
    ],
    "label": "otherOf",
  },
]
`;
